### ISSUE TITLE
Multi-GPU training has some Configuration error

### ISSUE THREAD

**Issue by johnjunjun7** [david8862/keras-YOLOv3-model-set#9]
I use this code :
`config = tf.ConfigProto()

config.gpu_options.allow_growth=True   #dynamic alloc GPU resource
config.gpu_options.per_process_gpu_memory_fraction = 0.99  #GPU memory threshold 0.3
session = tf.Session(config=config)`

then I set the batch_size=256, the use of GPU:
`| NVIDIA-SMI 418.39                       Driver Version: 418.39           CUDA Version: 10.1      |

| GPU Name           Persistence-M| Bus-Id                  Disp.A |       Volatile Uncorr. ECC |
| Fan Temp Perf Pwr:Usage/Cap   |               Memory-Usage |     GPU-Util Compute M. |

| N/A     46C      P0      83W / 300W | 31432MiB / 32480MiB |      70%               Default |

| N/A     32C      P0      67W / 300W |     4936MiB / 32480MiB |      100%          Default |

| N/A     35C      P0      68W / 300W |     4936MiB / 32480MiB |      25%          Default |

| N/A     35C      P0      73W / 300W |     4936MiB / 32480MiB |      100%          Default |`
I can't make full use of every GPU, because if I set the batch_size more than 256, The GPU give the OOM error;

I try to change the model from nanonet to **shufflenet_v2** with the **512batch_size**, it can train normally, and give this : 

`| NVIDIA-SMI 418.39                       Driver Version: 418.39           CUDA Version: 10.1      |

| GPU Name           Persistence-M| Bus-Id                  Disp.A |       Volatile Uncorr. ECC |
| Fan Temp Perf Pwr:Usage/Cap   |               Memory-Usage |     GPU-Util Compute M. |

| N/A     47C      P0     112W / 300W | 31424MiB / 32480MiB |          9%               Default |

| N/A     34C      P0     111W / 300W |     9032MiB / 32480MiB |      32%          Default |

| N/A     36C      P0      80W / 300W |     9032MiB / 32480MiB |      61%          Default |

| N/A     36C      P0      94W / 300W |     9032MiB / 32480MiB |      66%          Default |`

That makes me confused :( 
because the total number of parameters with **shufflenet_v2** is more than **nanonet.**

The reason why I try bigger batch_size just because when I set the batch_size to 256, this model just  achieve **74%** of top5, which is not good, so I try to increase the batch_size but failed. Maybe you can give me some suggestions. many thanks :)

**Comment by david8862** [comment]
Actually I didn't have much experience on multi-gpu training:) but there're some potential points which may impact the GPU memory usage in your case:

1. data parallelism strategy for keras multi_gpu_model (described [here](https://keras.io/zh/utils/#multi_gpu_model))
2. "max_queue_size" and "workers" value for data generator (described [here](https://keras.io/zh/models/model/#fit_generator))
3. OPs type in model (some tf OPs may only supported on CPU side)

You can try to adjust 1 and 2 for better parallelism performance. Hope that help :)

> I use this code :
> `config = tf.ConfigProto()
> 
> config.gpu_options.allow_growth=True #dynamic alloc GPU resource
> config.gpu_options.per_process_gpu_memory_fraction = 0.99 #GPU memory threshold 0.3
> session = tf.Session(config=config)`
> 
> then I set the batch_size=256, the use of GPU:
> `| NVIDIA-SMI 418.39 Driver Version: 418.39 CUDA Version: 10.1 |
> 
> | GPU Name Persistence-M| Bus-Id Disp.A | Volatile Uncorr. ECC |
> | Fan Temp Perf Pwr:Usage/Cap | Memory-Usage | GPU-Util Compute M. |
> 
> | N/A 46C P0 83W / 300W | 31432MiB / 32480MiB | 70% Default |
> 
> | N/A 32C P0 67W / 300W | 4936MiB / 32480MiB | 100% Default |
> 
> | N/A 35C P0 68W / 300W | 4936MiB / 32480MiB | 25% Default |
> 
> | N/A 35C P0 73W / 300W | 4936MiB / 32480MiB | 100% Default |`
> I can't make full use of every GPU, because if I set the batch_size more than 256, The GPU give the OOM error;
> 
> I try to change the model from nanonet to **shufflenet_v2** with the **512batch_size**, it can train normally, and give this :
> 
> `| NVIDIA-SMI 418.39 Driver Version: 418.39 CUDA Version: 10.1 |
> 
> | GPU Name Persistence-M| Bus-Id Disp.A | Volatile Uncorr. ECC |
> | Fan Temp Perf Pwr:Usage/Cap | Memory-Usage | GPU-Util Compute M. |
> 
> | N/A 47C P0 112W / 300W | 31424MiB / 32480MiB | 9% Default |
> 
> | N/A 34C P0 111W / 300W | 9032MiB / 32480MiB | 32% Default |
> 
> | N/A 36C P0 80W / 300W | 9032MiB / 32480MiB | 61% Default |
> 
> | N/A 36C P0 94W / 300W | 9032MiB / 32480MiB | 66% Default |`
> 
> That makes me confused :(
> because the total number of parameters with **shufflenet_v2** is more than **nanonet.**
> 
> The reason why I try bigger batch_size just because when I set the batch_size to 256, this model just achieve **74%** of top5, which is not good, so I try to increase the batch_size but failed. Maybe you can give me some suggestions. many thanks :)

**Comment by david8862** [comment]
And to improve accuracy on imagenet, you can also try to adjust the optimizer/learning_rate/decay policy. Although author didn't mention, but since the nanonet backbone is not designed for classification task, I guess the final top-1/top-5 accuracy should not reach SOTA level like MobileNet/ShuffleNet.

> I use this code :
> `config = tf.ConfigProto()
> 
> config.gpu_options.allow_growth=True #dynamic alloc GPU resource
> config.gpu_options.per_process_gpu_memory_fraction = 0.99 #GPU memory threshold 0.3
> session = tf.Session(config=config)`
> 
> then I set the batch_size=256, the use of GPU:
> `| NVIDIA-SMI 418.39 Driver Version: 418.39 CUDA Version: 10.1 |
> 
> | GPU Name Persistence-M| Bus-Id Disp.A | Volatile Uncorr. ECC |
> | Fan Temp Perf Pwr:Usage/Cap | Memory-Usage | GPU-Util Compute M. |
> 
> | N/A 46C P0 83W / 300W | 31432MiB / 32480MiB | 70% Default |
> 
> | N/A 32C P0 67W / 300W | 4936MiB / 32480MiB | 100% Default |
> 
> | N/A 35C P0 68W / 300W | 4936MiB / 32480MiB | 25% Default |
> 
> | N/A 35C P0 73W / 300W | 4936MiB / 32480MiB | 100% Default |`
> I can't make full use of every GPU, because if I set the batch_size more than 256, The GPU give the OOM error;
> 
> I try to change the model from nanonet to **shufflenet_v2** with the **512batch_size**, it can train normally, and give this :
> 
> `| NVIDIA-SMI 418.39 Driver Version: 418.39 CUDA Version: 10.1 |
> 
> | GPU Name Persistence-M| Bus-Id Disp.A | Volatile Uncorr. ECC |
> | Fan Temp Perf Pwr:Usage/Cap | Memory-Usage | GPU-Util Compute M. |
> 
> | N/A 47C P0 112W / 300W | 31424MiB / 32480MiB | 9% Default |
> 
> | N/A 34C P0 111W / 300W | 9032MiB / 32480MiB | 32% Default |
> 
> | N/A 36C P0 80W / 300W | 9032MiB / 32480MiB | 61% Default |
> 
> | N/A 36C P0 94W / 300W | 9032MiB / 32480MiB | 66% Default |`
> 
> That makes me confused :(
> because the total number of parameters with **shufflenet_v2** is more than **nanonet.**
> 
> The reason why I try bigger batch_size just because when I set the batch_size to 256, this model just achieve **74%** of top5, which is not good, so I try to increase the batch_size but failed. Maybe you can give me some suggestions. many thanks :)

**Comment by johnjunjun7** [comment]
Many thanks ：）
I want to try to improve the accuracy of yolonano through this pre-training method. Currently, the best mAP on VOC2007 is only 57%. 
I am also trying to test the performance of nano using the multi-scale training you just updated。
